File: Assets/Scripts/Core/Board.cs
using System.Collections.Generic;
using Sokoban.Levels;
using UnityEngine;

namespace Sokoban.Core
{
    /// <summary>Trạng thái một màn đang chơi. Lưới tĩnh không đổi; người chơi và hộp thì đổi.</summary>
    public class Board
    {
        readonly CellType[,] _statics;

        public int Width { get; }
        public int Height { get; }
        public Vector2Int PlayerPos { get; set; }
        public HashSet<Vector2Int> Boxes { get; }

        Board(int width, int height)
        {
            Width = width;
            Height = height;
            _statics = new CellType[width, height];
            Boxes = new HashSet<Vector2Int>();
        }

        /// <summary>Ngoài lưới coi như tường, nên nơi khác không cần kiểm tra biên.</summary>
        public CellType GetCell(Vector2Int p) =>
            p.x < 0 || p.x >= Width || p.y < 0 || p.y >= Height
                ? CellType.Wall
                : _statics[p.x, p.y];

        public bool HasBox(Vector2Int p) => Boxes.Contains(p);

        public bool IsSolved
        {
            get
            {
                foreach (var b in Boxes)
                    if (GetCell(b) != CellType.Goal) return false;
                return true;
            }
        }

        public static Board FromLevel(LevelData level)
        {
            var board = new Board(level.width, level.height);

            for (int y = 0; y < level.height; y++)
            {
                string row = level.rows[y];
                for (int x = 0; x < level.width; x++)
                {
                    char c = x < row.Length ? row[x] : SokobanChars.Floor;
                    var pos = new Vector2Int(x, y);

                    board._statics[x, y] = c == SokobanChars.Wall ? CellType.Wall
                        : c == SokobanChars.Goal || c == SokobanChars.BoxOnGoal ||
                          c == SokobanChars.PlayerOnGoal ? CellType.Goal
                        : CellType.Floor;

                    if (c == SokobanChars.Player || c == SokobanChars.PlayerOnGoal)
                        board.PlayerPos = pos;
                    if (c == SokobanChars.Box || c == SokobanChars.BoxOnGoal)
                        board.Boxes.Add(pos);
                }
            }

            return board;
        }
    }
}

File: web/src/core/board.js
import {
  WALL, FLOOR, GOAL, BOX, BOX_ON_GOAL, PLAYER, PLAYER_ON_GOAL,
} from '../levels/sokobanChars.js';

export const CellType = Object.freeze({
  Wall: 'Wall',
  Floor: 'Floor',
  Goal: 'Goal',
});

/**
 * Boxes live in a Set as "x,y" string keys. JS compares objects by reference, so a
 * Set of {x,y} would never recognise two equal coordinates — a forced difference
 * from the HashSet<Vector2Int> the C# original used.
 */
export const boxKey = (x, y) => `${x},${y}`;

export function parseBoxKey(key) {
  const [x, y] = key.split(',').map(Number);
  return { x, y };
}

/** State of a level in play. The static grid never changes; the player and boxes do. */
export class Board {
  constructor(width, height) {
    this.width = width;
    this.height = height;
    // statics[y][x] — same order as LevelData.rows, so reading the code needs no mental flip.
    this.statics = Array.from({ length: height }, () => new Array(width).fill(CellType.Floor));
    this.player = { x: 0, y: 0 };
    this.boxes = new Set();
  }

  /** Outside the grid counts as wall, so nowhere else has to bounds-check. */
  cellAt(x, y) {
    if (x < 0 || x >= this.width || y < 0 || y >= this.height) return CellType.Wall;
    return this.statics[y][x];
  }

  hasBox(x, y) {
    return this.boxes.has(boxKey(x, y));
  }

  get isSolved() {
    for (const key of this.boxes) {
      const { x, y } = parseBoxKey(key);
      if (this.cellAt(x, y) !== CellType.Goal) return false;
    }
    return true;
  }

  static fromLevel(level) {
    const board = new Board(level.width, level.height);

    for (let y = 0; y < level.height; y++) {
      const row = level.rows[y];
      for (let x = 0; x < level.width; x++) {
        // Rows shorter than width: the missing part is empty floor.
        const c = x < row.length ? row[x] : FLOOR;

        if (c === WALL) board.statics[y][x] = CellType.Wall;
        else if (c === GOAL || c === BOX_ON_GOAL || c === PLAYER_ON_GOAL) board.statics[y][x] = CellType.Goal;
        else board.statics[y][x] = CellType.Floor;

        if (c === PLAYER || c === PLAYER_ON_GOAL) board.player = { x, y };
        if (c === BOX || c === BOX_ON_GOAL) board.boxes.add(boxKey(x, y));
      }
    }

    return board;
  }
}
